default session saving completed & tested

gigajuwels opened the pull request.

gigajuwels committed:
feat: default session saving completed & tested

gigajuwels committed:
fix: auth client regression

gigajuwels committed:
fix: print stored keys, passes test

skibitsky commented on Core Modules/WalletConnectSharp.Storage/InMemoryStorage.cs:
Why not use `ConcurrentDictionary` for `Entries` instead of `Dictionary` + `lock` statements?

https://learn.microsoft.com/en-us/dotnet/standard/collections/thread-safe/when-to-use-a-thread-safe-collection#concurrentdictionary-vs-dictionary

skibitsky commented on Tests/WalletConnectSharp.Sign.Test/SignTests.cs:
```diff
@@ -384,50 +389,50 @@ public async void TestTwoUniqueSessionRequestResponse()
         
         [Fact, Trait("Category", "integration")]
-        public async void TestTwoUniqueSessionRequestResponseUsingAddressProviderDefaults()
+        public async Task TestTwoUniqueSessionRequestResponseUsingAddressProviderDefaults()
```
Is there a specific reason why most test methods return void instead of Task? If not, I recommend switching to Task to follow best practices.

skibitsky commented on WalletConnectSharp.Core/Controllers/Relayer.cs:
```diff
@@ -199,33 +199,46 @@ protected virtual Task<IJsonRpcConnection> BuildConnection(string url)
-            {
-                OnProviderPayload(s);
-            };
+            Provider.RawMessageReceived += ProviderOnRawMessageReceived;
+            Provider.Connected += ProviderOnConnected;
+            Provider.Disconnected += ProviderOnDisconnected;
+            Provider.ErrorReceived += ProviderOnErrorReceived;
```
Let's try to be consistent with naming when possible. 

`ProviderOnRawMessageReceived` →  `OnProviderRawMessageReceived`

skibitsky commented on WalletConnectSharp.Core/Models/MessageHandler/TypedEventHandler.cs:
```diff
@@ -258,5 +258,11 @@ async Task<VerifiedContext> VerifyContext(string hash, Metadata metadata)
 
             return context;
         }
+
+        public virtual void Dispose()
+        {
+            var context = _ref.Context;
+            _instances.Remove(context);
+        }
```
Please implement disposal with `Dispose(bool)` overload

https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#dispose-and-disposebool

skibitsky commented on WalletConnectSharp.Sign/Models/DefaultsLoadingEventArgs.cs:
```diff
@@ -0,0 +1,13 @@
+
+public class DefaultsLoadingEventArgs : EventArgs
+{
+    public DefaultsLoadingEventArgs(ref AddressProvider.DefaultData defaults)
```
`ref` →  `in`

skibitsky commented on WalletConnectSharp.Sign/Controllers/AddressProvider.cs:
```diff
@@ -46,7 +91,27 @@ public AddressProvider(ISignClient client)
+        var key = $"{Context}-default-session";
+        if (await _client.Core.Storage.HasItem(key))
+        {
+            _state = await _client.Core.Storage.GetItem<DefaultData>($"{Context}-default-session");
```
Use `key` variable from L104?

skibitsky commented on WalletConnectSharp.Sign/Controllers/AddressProvider.cs:
```diff
@@ -75,45 +140,52 @@ private void ClientOnSessionConnected(object sender, SessionStruct e)
+        finally
         {
-            DefaultNamespace = null;
+            await SaveDefaults();
```
Why is this in `finally`? What if `try` block breaks halfway and we save incorrect session?

skibitsky commented on WalletConnectSharp.Sign/Controllers/AddressProvider.cs:
```diff
@@ -75,45 +140,52 @@ private void ClientOnSessionConnected(object sender, SessionStruct e)
+    private async void UpdateDefaultChainAndNamespace()
     {
-        if (HasDefaultSession)
+        try
```
Let's add `catch` block to log exception with `WCLogger`.

skibitsky commented on WalletConnectSharp.Sign/Controllers/AddressProvider.cs:
```diff
@@ -1,11 +1,22 @@
+        public string ChainId;
+    }
+
+    public event EventHandler<DefaultsLoadingEventArgs> DefaultsLoading;
```
Why is it called DefaultsLoad**ing**? If I'm reading the code correctly, the event is invoked after the default session state is loaded.

skibitsky commented on WalletConnectSharp.Sign/Models/SessionRequestEventHandler.cs:
```diff
@@ -44,20 +45,32 @@ protected SessionRequestEventHandler(ICore engine, IEnginePrivate enginePrivate)
+            _disposeActions.Add(() =>
+            {
+                instance.Dispose();
+            });
```
Here and in the other places it's possible to use method group instead of lambda
```csharp
_disposeActions.Add(wrappedRef.Dispose);
```

skibitsky commented on WalletConnectSharp.Sign/Models/SessionRequestEventHandler.cs:
```diff
@@ -94,5 +107,15 @@ await _enginePrivate.SetPendingSessionRequest(new PendingRequestStruct()
+            }
+            
+            base.Dispose();
+        }
```
Please implement disposal with `Dispose(bool)` overload

https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#dispose-and-disposebool

skibitsky reviewed (changes requested):
Good! Just a few small requests and questions

skibitsky commented:
There is another issue that can break session requests.

Engine now uses `OrderedKeys` field when making requests:
https://github.com/WalletConnect/WalletConnectSharp/blob/feat/default-session-saving/WalletConnectSharp.Sign/Engine.cs#L687

The problem I'm sometimes facing is that `RequiredNamespace` object can have one `Key`, but no `OrderedKeys`. Perhaps there is a scenario when dictionary is assigned directly, bypassing the `Add` method of `RequiredNamespaces`.

gigajuwels committed:
fix: dispose pattern, test return type and other

gigajuwels committed:
fix: use default namespace and chain in Engine

gigajuwels committed:
Merge branch '2.0' into feat/default-session-saving

gigajuwels committed:
fix: compile errors

gigajuwels committed:
chore: rename listener functions to match naming

skibitsky commented:
@gigajuwels the `TestAddressProviderDefaultsSaving` test fails

gigajuwels committed:
fix: dispose loop

gigajuwels committed:
feat: AddressProvider uses Namespaces instead of RequiredNamespaces
also introduce SortedDictionary

gigajuwels committed:
feat: added relay batch message fetching and fixed several lifecycle issues

gigajuwels committed:
test: Ensure namespaces include Chains array

gigajuwels committed:
chore: remove debug logging

gigajuwels committed:
chore: replace foreach await with task.whenall

skibitsky committed:
Make udpated FileSystemStorage backwards compatibile

skibitsky committed:
Fix namespaces not being deserialized correctly

`OrderedKeys` has always been 0 when deserializing the Namespaces JSON. The issue was solved by replacing `SortedDictionary` with an implementation from `System.Collections.Generic`.

skibitsky committed:
Merge branch '2.0' into feat/default-session-saving

# Conflicts:
#	Core Modules/WalletConnectSharp.Storage/FileSystemStorage.cs

skibitsky commented:
Well done, @gigajuwels!
